modules/system/files/linker.py

```diff
@@ -0,0 +1,150 @@
+            else:
+                # The old generation did not have a file at this path,
+                # and we never overwrite links that weren't created by us
+                problems.append(path)
+        else:
+            # The file is a regular file
+            problems.append(path)
+    else:
+        # There is no file at this path
+        transactions.append(Transaction(new_file['source'], path, new_file['type']))
+
+# Check problems
+for problem in problems:
+    print(f"Existing file at path {problem}")
+
+if len(problems) > 0:
+    print("Aborting")
+    exit(1)
+
+if CHECK_ONLY:
+    # We don't perform any checks when planning removal of old files, so we can exit here
+    exit(0)
+
+# Remove all remaining files from the old generation that aren't in the new generation
+path: str
+for path in old_files['files']:
+    old_file = old_files['files'][path]
+    if path in new_files['files']:
+        # Already handled when we iterated through new_files above
+        continue
+    if os.path.lexists(path):
+        # There is a file at this path
+        # It could be a regular file or a symlink (including broken symlinks)
+
+        if os.path.islink(path):
+            # The file is a symlink
+
+            if old_file['type'] != "link":
+                # This files wasn't a link at last activation, which means that the user changed it
+                # Therefore we don't touch it
+                continue
+
+            # Check that its destination remains the same
+            link_target = os.readlink(path)
+            if os.path.join(os.path.dirname(path), link_target) == old_file['source']:
+                # The link has not changed since last system activation, so we can overwrite it
+                transactions.append(Transaction(path, path, "remove"))
+            else:
+                # The link is to somewhere else, so leave it alone
+                continue
+        else:
+            # The file is a regular file
+            continue
+    else:
+        # There's no file at this path anymore, so we have nothing to do anyway
+        continue
+
+# Perform all transactions
+for t in transactions:
+    # NOTE: the naming scheme for transaction properties is confusing
+    # We are **NOT** using the same scheme as symlinks when we talk about
+    # source/destination. The way we are using these names, `source` is a path
+    # in the Nix store, and `destination` is the path in the system where the source
+    # should be linked or copied to.
+    # In the special case of removing files, `destination` can be ignored
+    if DRY_RUN:
+        match t.type:
+            case "link":
+                print(f"ln -s {t.source} {t.destination}")
+            case "remove":
+                print(f"rm {t.source}")
+            case _:
+                print(f"Unknown transaction type {t.type}")
+    else:
+        match t.type:
+            case "link":
+                # TODO: ensure enclosing directory exists
+
```

Comment:
The one outstanding change for this PR is ensuring that the enclosing directories for new links exist. Copying something from Matrix:

> - Creating enclosing directories when files are linked into place originally: I haven't done this yet, because I wasn't sure if we just want to ensure that enclosing directories exist or also do something special to keep track of the fact that we made them so that we can remove them when necessary. If we do want to track this, how so? If we don't, see the next point:
> - Removing directories when files in them are removed: do we do this at all? how do we know when to do so? My current thought is that we only remove the actual files themselves, *but* if we remove the last file in a directory we also print a message during activation. I think this is reasonable because you'll only get the message as you switch from a generation that does have the nested file to a generation that doesn't have the nested file.

Reply:
There's also a third option, which is to leave intermediate directories but print a warning whenever the actions taken from activation result in an empty directory. I think this is the best option